p2p/ipld/plugin/plugin_test.go

```diff
@@ -0,0 +1,222 @@
+			}
+			nodePrefixOffset := 1 // leaf / inner node prefix is one byte
+			lastLeafNodeData := gotNodes[len(gotNodes)-1].RawData()
+			if gotData, wantData := lastLeafNodeData[nodePrefixOffset:], tt.leafData[0]; !bytes.Equal(gotData, wantData) {
+				t.Errorf("first node's data does not match the leaf's data\ngot: %v\nwant: %v", gotData, wantData)
+			}
+		})
+	}
+}
+
+func TestDagPutWithPlugin(t *testing.T) {
+	t.Skip("Requires running ipfs daemon with the plugin compiled and installed")
+
+	t.Log("Warning: running this test writes to your local IPFS block store!")
+
+	const numLeaves = 32
+	data := generateRandNamespacedRawData(numLeaves, namespaceSize, shareSize)
+	buf := createByteBufFromRawData(t, data)
+	printFirst := 10
+	t.Logf("first leaf, nid: %x, data: %x...", data[0][:namespaceSize], data[0][namespaceSize:namespaceSize+printFirst])
+	n := nmt.New(sha256.New())
+	for _, share := range data {
+		err := n.Push(share[:namespaceSize], share[namespaceSize:])
+		if err != nil {
+			t.Errorf("nmt.Push() unexpected error = %v", err)
+			return
+		}
+	}
+
+	sh := shell.NewLocalShell()
+	cid, err := sh.DagPut(buf, "raw", DagParserFormatName)
+	if err != nil {
+		t.Fatalf("DagPut() failed: %v", err)
+	}
+	// convert NMT tree root to CID and verify it matches the CID returned by DagPut
+	treeRootBytes := n.Root().Bytes()
+	nmtCid, err := cidFromNamespacedSha256(treeRootBytes)
+	if err != nil {
+		t.Fatalf("cidFromNamespacedSha256() failed: %v", err)
+	}
+	if nmtCid.String() != cid {
+		t.Errorf("CIDs from NMT and plugin do not match: got %v, want: %v", cid, nmtCid.String())
+	}
+	// print out cid s.t. it can be used on the commandline
+	t.Logf("Stored with cid: %v\n", cid)
+
+	// DagGet leaf by leaf:
+	for i, wantShare := range data {
+		gotLeaf := &nmtLeafNode{}
+		path := leafIdxToPath(cid, i)
+		if err := sh.DagGet(path, gotLeaf); err != nil {
+			t.Errorf("DagGet(%s) failed: %v", path, err)
+		}
+		if gotShare := gotLeaf.Data; !bytes.Equal(gotShare, wantShare) {
+			t.Errorf("DagGet returned different data than pushed, got: %v, want: %v", gotShare, wantShare)
+		}
+	}
+}
+
+func generateExtendedRow(t *testing.T) [][]byte {
+	origData := generateRandNamespacedRawData(16, namespaceSize, shareSize)
+	origDataWithoutNamespaces := make([][]byte, 16)
+	for i, share := range origData {
+		origDataWithoutNamespaces[i] = share[namespaceSize:]
+	}
+
+	if extendedData, err := rsmt2d.ComputeExtendedDataSquare(origDataWithoutNamespaces, rsmt2d.RSGF8, newNmtConstructor); err != nil {
+		t.Fatalf("rsmt2d.Encode(): %v", err)
+		return nil
+	} else {
+		extendedRow := extendedData.Row(0)
+		for i, rowCell := range extendedRow {
+			if i < len(origData)/4 {
+				nid := origData[i][:namespaceSize]
+				extendedRow[i] = append(nid, rowCell...)
+			} else {
+				maxNid := bytes.Repeat([]byte{0xFF}, namespaceSize)
+				extendedRow[i] = append(maxNid, rowCell...)
+			}
+		}
```

Comment:
```suggestion
	extendedData, err := rsmt2d.ComputeExtendedDataSquare(origDataWithoutNamespaces, rsmt2d.RSGF8, newNmtConstructor)
	if err != nil {
		t.Fatalf("rsmt2d.Encode(): %v", err)
		return nil
	}
	extendedRow := extendedData.Row(0)
	for i, rowCell := range extendedRow {
		if i < len(origData)/4 {
			nid := origData[i][:namespaceSize]
			extendedRow[i] = append(nid, rowCell...)
		} else {
			maxNid := bytes.Repeat([]byte{0xFF}, namespaceSize)
			extendedRow[i] = append(maxNid, rowCell...)
		}
	}
	return extendedRow
}
```
This is strictly personal opinion. I think there might be a tad too much going on in the first if statement, so using the slightly more typical format might improve readability. Certainly not a must-change, though.